src/app/shared/filesystem/DirectoryTree.ts

```diff
@@ -382,17 +314,29 @@ export class DirectoryNode {
    * @memberof DirectoryNode
    */
   public addFile(newFile: LearningObject.Material.File) {
-    let canAdd = true;
-    for (const file of this.files) {
-      if (file.name === newFile.name) {
-        canAdd = false;
-        break;
-      }
-    }
-    if (canAdd) {
+    if (this.fileMap.get(newFile.name) == null) {
+      this.fileMap.set(newFile.name, this.files.length);
       this.files.push(newFile);
     }
   }
+  /**
+   * Remove folder from Node's children by folderName
+   *
+   * @param {string} folderName
+   * @returns {DirectoryNode}
+   * @memberof DirectoryNode
+   */
+  public removeFolder(folderName: string): DirectoryNode {
+    const index = this.folderMap.get(folderName);
+    if (index >= 0) {
+      const deleted = this.folders[index];
+      this.folders.splice(index, 1);
+      this.folderMap.delete(folderName);
+      this.reIndexCacheMap(index, this.folders, this.folderMap);
+      return deleted;
+    }
+    return null;
```

Comment:
Question on the return `null`'s: Why did you choose to explicitly return null, compared to just letting undefined come back as default?

Reply:
@DonSeannelly Since the return type is defined all execution branches must return.